src/secret_sharing/array.rs

```diff
@@ -0,0 +1,245 @@
+use std::{
+    array,
+    fmt::Debug,
+    ops::{Add, AddAssign, Mul, Neg, Sub, SubAssign},
+};
+
+use generic_array::{ArrayLength, GenericArray};
+use typenum::U32;
+
+use crate::{
+    ff::{Field, Serializable},
+    helpers::Message,
+    secret_sharing::{SharedValue, SharedValueArray},
+};
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct StdArray<V, const N: usize>([V; N]);
+
+// Introducing the `Eq` bound here is a wart. For some reason AdditiveShare is `Eq`, but
+// `SharedValue` is not. Probably all of these should be Eq, or none of them.
+impl<V: SharedValue + Eq, const N: usize> SharedValueArray<V> for StdArray<V, N> {
+    const ZERO: Self = Self([V::ZERO; N]);
+
+    fn capacity() -> usize {
+        N
+    }
+
+    fn index(&self, index: usize) -> V {
+        self.0[index]
+    }
+
+    fn from_item(item: V) -> Self {
+        let mut res = Self::ZERO;
+        res.0[0] = item;
+        res
+    }
+
+    /*
+    fn multiply_elements(lhs: &Self, rhs: &Self) -> Self {
+        StdArray(array::from_fn(|i| lhs.0[i] * rhs.0[i]))
+    }
+    */
+}
+
+impl<V: SharedValue, const N: usize> TryFrom<Vec<V>> for StdArray<V, N> {
+    type Error = ();
+    fn try_from(value: Vec<V>) -> Result<Self, Self::Error> {
+        value.try_into().map(Self).map_err(|_| ())
+    }
+}
+
+impl<'a, 'b, V: SharedValue, const N: usize> Add<&'b StdArray<V, N>> for &'a StdArray<V, N> {
+    type Output = StdArray<V, N>;
+
+    fn add(self, rhs: &'b StdArray<V, N>) -> Self::Output {
+        //StdArray(self.0.iter().zip(rhs.0.iter()).map(|(a, b)| *a + *b).collect::<Vec<_>>().try_into().unwrap())
+        StdArray(array::from_fn(|i| self.0[i] + rhs.0[i]))
```

Comment:
This has the same role as the corresponding operation implemented on the BA* types. The BA types implement elementwise operations on an array of bits using bitwise AND, XOR. This is an implementation of elementwise operations on an array of Fp32BitPrime or Fp25519 or similar.

Reply:
I assumed that we would pass an array without boolean operations to a function and if we need to add within the function, we would implement this explicitly. I am wondering, if I e.g. add x,y,z together, this would need two iteration. Would a compiler be able to optimize this and merge it into a single iteration? At least for Fp25519 types, there are actually not that many places where we need basic operations like additions. However there are other operations like mapping them to curve points where we also would need to explicitly iterate over the vector within a the function.

Reply:
related, does the compiler currently unroll all the loops and rearrange assembly instructions for cache optimization?

Reply:
Yes, the compiler will unroll these, map them to SIMD intrinsics, etc.